runtime/vcache/cache.go

```diff
@@ -19,27 +19,50 @@ type Cache struct {
 	// files and close them and then reopen them when needed to access
 	// vectors that haven't yet been loaded.
 	objects map[ksuid.KSUID]*Object
+	locks   map[ksuid.KSUID]*sync.Mutex
 }
 
 func NewCache(engine storage.Engine) *Cache {
 	return &Cache{
 		engine:  engine,
 		objects: make(map[ksuid.KSUID]*Object),
+		locks:   make(map[ksuid.KSUID]*sync.Mutex),
 	}
 }
 
-func (c *Cache) Fetch(ctx context.Context, uri *storage.URI, id ksuid.KSUID) (*Object, error) {
-	//XXX do we want finer grained mutex?  might be ok if lookups always done inside
-	// a dedicated goroutine
+func (c *Cache) Lock(id ksuid.KSUID) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
+	mu, ok := c.locks[id]
+	if !ok {
+		mu = &sync.Mutex{}
+		c.locks[id] = mu
+	}
+	mu.Lock()
+}
+
+func (c *Cache) Unlock(id ksuid.KSUID) {
+	c.mu.Lock()
+	c.locks[id].Unlock()
+	c.mu.Unlock()
+}
+
+func (c *Cache) Fetch(ctx context.Context, uri *storage.URI, id ksuid.KSUID) (*Object, error) {
+	c.mu.Lock()
 	if object, ok := c.objects[id]; ok {
+		c.mu.Unlock()
 		return object, nil
 	}
-	object, err := NewObject(ctx, c.engine, uri, id)
+	c.mu.Unlock()
+	c.Lock(id)
+	object, err := NewObject(ctx, c.engine, uri)
 	if err != nil {
+		c.Unlock(id)
 		return nil, err
 	}
+	c.mu.Lock()
 	c.objects[id] = object
+	c.mu.Unlock()
+	c.Unlock(id)
```

Comment:
If you and I call `c.Lock(id)` concurrently and I block waiting for you, then you'll set `c.object[id]` to the object you've created and, when you call `c.Unlock(id)`, I'll overwrite that with the object I've created.

I'm not sure if that's OK, but if it isn't, we could do this.
```suggestion
	object, err := NewObject(ctx, c.engine, uri)
	if err != nil {
		return nil, err
	}
	c.mu.Lock()
	if o, ok := c.objects[id]; ok {
		object = o
	} else {
		c.objects[id] = object
	}
	c.mu.Unlock()
```

Reply:
It doesn't seem right to created an object that's not used